File: lib/faye/util/namespace.rb
module Faye
  class Namespace

    extend Forwardable
    def_delegator :@used, :delete, :release
    def_delegator :@used, :has_key?, :exists?

    def initialize
      @used = {}
    end

    def generate
      name = Engine.random
      name = Engine.random while @used.has_key?(name)
      @used[name] = name
    end

  end
end

File: javascript/util/namespace.js
Faye.Namespace = Faye.Class({
  initialize: function() {
    this._used = {};
  },

  exists: function(id) {
    return this._used.hasOwnProperty(id);
  },

  generate: function() {
    var name = Faye.random();
    while (this._used.hasOwnProperty(name))
      name = Faye.random();
    return this._used[name] = name;
  },

  release: function(id) {
    delete this._used[id];
  }
});
